### ISSUE TITLE
Compatibility with IE11 (Internet Explorer 11)

### ISSUE THREAD

**Issue by maxbarnas** [ckeditor/ckeditor5#330]
Long story short – CKEditor 5 doesn't work on IE11. Making it compatible with IE11 is a complex and long task – [read more about the status of things and how you can help](https://github.com/ckeditor/ckeditor5/issues/330#issuecomment-438016342).

**Comment by Reinmar** [comment]
Super, thanks. It's usually a good news if many failed tests share the same error :D.

**Comment by djanosik** [comment]
There is a problem either in transpiler or in polyfill related to `Symbol` which causes [this condition](https://github.com/ckeditor/ckeditor5-utils/blob/v0.9.1/src/observablemixin.js#L276) to be always `true`. The attributes property will never get initialized and the TypeError is thrown. When I fix this, the editor starts to crash (or hang) in IE11.

**Comment by Reinmar** [comment]
Interesting... this doesn't seem to be some tricky scenario so it's odd that Babel fails on that. How did you change it so it worked? `observable[ attributeSymbol ]`?

**Comment by Reinmar** [comment]
Nah, `observable[ attributeSymbol ]` would be unsafe. `observable.hasOwnProperty( attributeSymbol )`?

**Comment by djanosik** [comment]
@Reinmar Yes, `hasOwnProperty` works quite well.

**Comment by Reinmar** [comment]
So, I guess the next step would be to run `ckeditor5-utils` tests and then `ckeditor5-engine`. That may give some idea what's crashing. But since this requires transpiling the code we'd need support for that in https://github.com/ckeditor/ckeditor5-dev/blob/master/packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js (we could add it under `--es5` flag... I even think we supported it at some stage :D)

**Comment by djanosik** [comment]
Some other issues:
- Error on `document.createTreeWalker` missing some args.
- Crashes and hangs on [container.remove()](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/renderer.js#L639), because `remove()` function doesn't exists and is being called many (like hundred) times. There is probably some issue with callbacks.
- `Selection.extend(...)` is missing. When I comment out [this line](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/renderer.js#L607) it starts to work, but selection is not reliable.

**Comment by Reinmar** [comment]
Yeah, good old browsers :|. 

I think that before we'll start fixing these issue we need to figure out how to do that without bloating the existing code. By the issues you found it seems that we need more polyfills for missing/outdated methods. I wonder if we can safely polyfill all these.

**Comment by djanosik** [comment]
Any chance first release of CKE5 will work in IE11?

**Comment by Reinmar** [comment]
It depends. We don't know how far we are from making CKE5 really work in IE11. We'll check that after releasing 1.0.0 alpha and see if we'll be able to bring IE11 support for 1.0.0 final.

**Comment by djanosik** [comment]
Ok, thanks. Our decision whether to use CKE5 depends on it. Unfortunately we have many customers who are still using IE11 :/

**Comment by wwalc** [comment]
I did a quick search and it looks like IE11 will be supported by MS for some good time still (as long as Win 10 is supported): https://support.microsoft.com/en-us/help/17454/lifecycle-faq-internet-explorer
... which means at least until the end of 2020 (?) https://support.microsoft.com/en-us/help/13853/windows-lifecycle-fact-sheet

**Comment by lindenthal** [comment]
Dear CK 5 team,

first of all thanks a lot for putting so much effort into this editor. Here at [OpenProject](https://www.openproject.org) we just did a spike with it. The development of custom plugins is so much easier when compared with other editors. So my congratulations to your design decisions. Your editor is really awesome!

Getting back to the IE 11 topic:

Here at OpenProject we had a very similar discussion 2 years ago. We decided to drop support for IE 11. We had to make a decision whether we want to invest our time into the **future** or in the **past**. And we decided to invest into the future. Looking back I think it was the right decision. The browser landscape has changed dramatically in the last two years. We work for a large number of large corporates and governmental organizations. All of them are allowed to use modern browsers now. So if you ask them: _Do we need to support IE 11_ they all say _Yes_. If you ask them _Are you allowed to use modern browsers_ they als say _Yes_. And if you ask them _Do you want new features_ they also say _Yes_. And if you ask them _Are you willing to pay the effort required to support IE 11_ they say **_No_**.

These people are slowing down the entire innovation by sticking to their legacy world. If application developers continue to support IE11 this will give them less pressure to move on.

In my opinion more people would benefit from features such as table-support. Or legal requirements such as full accessibility.

@wwalc and @Reinmar and @djanosik: wouldn't it be an option to provide (paid) long-term support for CK4 until the end of IE11 min 2020? And everybody who wants the new features of CK5 needs modern browsers?

Again: Thanks a lot for all the hard work you for the community. I really appreciate that.

**Comment by djanosik** [comment]
@lindenthal I guess it would be ok to use CKE4. But how do I create headless CKE4 editor? CKE5 is modular and replacing UI with our own is quite easy.

- _Do we need to support IE11?_ Yes :/
- _Are you allowed to use modern browsers?_ Yes, mostly. But our solutions are also integrated into legacy systems that don't work in modern browsers. Also don't forget IE11 is usually the only option available on Windows Servers. There is no Edge and admins don't allow to install 3rd party apps.
- _Do you want new features?_ Of course. But I think we can find a compromise here. We need CKE5 to not crash in IE11 and to support at least some basic editing features. I can explain why some features are not available. But I can't explain why it doesn't work at all. 
- _Are you willing to pay the effort required to support IE11?_ Not sure here.

**Comment by lindenthal** [comment]
Hi @djanosik,

I do understand your concerns but please allow me to exaggerate a little so you can better understand my position.

Actually you just confirmed my thesis: The development of modern web application is slown down by poorly maintained legacy applications. So people like the CKE developers, you and I have to suffer from that. And as I said: These companies that built up this technical debt try to hand over their problem to others. Which actually means that these companies are asking other open source developers to pay for their technical debt. 

But I guess there are many other opinions on that :-)

Best
Niels

**Comment by caffodian** [comment]
I work in a field where IE11 is unfortunately still required, and customers can't easily upgrade.  (Hospitals and other health) I haven't looked very much into what is actually broken with CKE5 under IE11, but I've had to deal with similar problems with polyfilling for IE11 in the past.

There are a few conditionals, but supporting IE11 (through webpack magic) would probably be something I'd be willing to pay for, or to help look into and work on.

**Comment by lindenthal** [comment]
@caffodian 

I suggest you first upgrade to modern browsers and then upgrade to CK5. What is wrong with that approach? I have never heard anybody who said they can not use a modern Edge, Chrome or FF because of stability or security concerns. Do you?

**Comment by code-chris** [comment]
@lindenthal 
Yes, I have. Our software is used from bigger companies. I've tried to stop IE support, but they said it's the default browser. There are many computers which do not run with Win 10 yet, so no Edge is available. And other browsers are not really wanted... I'm glad that we ONLY have to support IE11 and not older ones. IE century isn't over (especially for companies)

**Comment by lindenthal** [comment]
@code-chris 

I guess you agree that FF and Chrome both work perfectly fine on Windows 7, right? 
And even if the IE11 is the default browser in an organiatzion this does not imply that they can not use FF in parallel. And for the very small fraction that can not install FF, they should simply stick to CK4. 

The only reason to chose IE11 over FF are poorly maintained legacy web applications. I am not aware of any other reason.

The point is: The more applications continue to support IE 11 the longer it will take to get rid of it. So by supporting IE 11 we make this problem even worse.

You know there is another group of people that makes me even more angry: The people that request to support the combination of legacy browser with legacy screen readers (IE11 with JAWS 15). This combination really kills any UX and innovation. Did you ever see some of these people paying for the additional effort they create?

**Comment by code-chris** [comment]
@lindenthal 
I agree with you. No problem. I would like to kill IE immediately too. But unfortunately it's hard to stop IE support in commercial used software. And yes, as longer IE is supported from components like this one, as longer it will take until this browser dies completely.

The reason why I hope that CK5 will support IE is not cause of the browser:
- Our main product uses now a self-written Editor component which is really worse. I want to replace it with CK. But if I replace it with CK4, then I finally have duplicate effort (Self-Written to CK4 and later to CK5)
- CK4 does not support Webpack or has any other compatible solution.

**Comment by hubertguillemain** [comment]
Until the argument wether IE 11 shall or shall not be officially supported is closed, could anyone come up with a workaround solution existing using external polyfill libraries?

**Comment by Infuser** [comment]
I think anyone thinking IE11 support is not vital is not being realistic, I work for a large company that makes software for our customers a large number of them want IE11 support, I wish they didn’t but they do.

Currently I am developing a completely new front end in which I would like to use ckeditor5 but I cannot because of no IE11 support. Additionally it sounds like the upgrade path from ckeditor4 to 5 is difficult and we will have issues converting our customers old data. This means not only do we have to use ckeditor4 but we will be stuck on ckeditor4 forever.

**Comment by Reinmar** [comment]
To make it clear – while none of us personally want to work on adding support for IE11 (which feeling, I can see, you understand), it’s basically a matter of **funding and opportunities** for the project. You can always contact us (https://cksource.com/contact/) to discuss the options.

---

I've missed this question previously:

> @wwalc and @Reinmar and @djanosik: wouldn't it be an option to provide (paid) long-term support for CK4 until the end of IE11 min 2020? And everybody who wants the new features of CK5 needs modern browsers?

The answer can be found in our [Help center](https://support.ckeditor.com/hc/en-us) – ["How long will CKEditor 4 be supported?](https://support.ckeditor.com/hc/en-us/articles/115005281629-How-long-will-CKEditor-4-be-supported-)

> We are determined to continue the CKEditor 4 development and maintenance for some good time still.
>
>The CKEditor 4 roadmap looks as follow:
> 
> * Bug fixes and browser support - provided until January 2020.
> * Security bug fixes and an extended SLA support - provided until January 2023.

**Comment by Reinmar** [comment]
@long-lazuli asked a few questions how to start investigating IE11 support:

> I saw that main problems are es6-class & es6 getter/setter
> so I thought it can be possible to compile it with some babel plugins.

Please take a look at https://docs.ckeditor.com/ckeditor5/latest/builds/guides/integration/advanced-setup.html#option-building-to-es5-target. I hope it will help.

> then I start digging in babel-plugins compiling to see if I can make things work on old projects. the best I can imagine is to build a polyfill file dedicated to this browser.

It depends and, TBH, it's hard to tell without digging. The first thing to do will be to:

1. Make sure that you build CKEditor 5 to ES5, so there are no syntactic errors.
2. Make sure you load polyfills for ES's globals such as `Map/Set/WeakMap/WeakSet/Symbol`. 
3. Check out missing HTML globals and features. Here, it becomes slippery. Sometimes you may try to load some well-known polyfills. E.g. [`Node#remove()`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/remove) may require one. But some HTML features that we rely on may not have proper polyfills. It requires a lot of testing basically. @djanosik already did some research in https://github.com/ckeditor/ckeditor5/issues/330#issuecomment-311895564.
4. Finally, there are browser bugs, "incomplete implementations" and "it works like this cause why not; ouch, and this time it worked differently cause it can!". In the world of contentEditable that's pretty broken even today. In IE11 you'll have to chase issues one by one and in case of them polyfills won't help. You may need to propose PRs for them with `if ( iAmIE ) { doWeirdStuff() } else { doNormalStuff() }` pieces.

The first two steps are provided by Babel polyfills. Then, there's a matter of finding or writing polyfills for HTML features we rely on or finding workarounds to not need them at all (will require PRs). Then, it'll get even darker and more smelly and you'll be chasing bugs and quirks of which it's plenty in IE11. At this stage, though, the editor should be quite usable already.

Anyway... good luck :)

PS. The best way to share your work and findings will be:

1. To fork https://github.com/ckeditor/ckeditor5 and one of the builds repo.
2. Modify `ckeditor5/mgit.json` to use your build's fork.
3. Modify that build's `webpack.config.js` and whatever else you'll need to load all the polyfills and stuff required by steps 1-3.

Thanks to that, the next person, or we, will be able to jump in and help.

You can read more about out development env in https://docs.ckeditor.com/ckeditor5/latest/framework/guides/contributing/development-environment.html

**Comment by brianteeman** [comment]
Great news

**Comment by Reinmar** [comment]
> I've got something together that seems to be working so far.

🎉 awesome :)

> I have made a separate local repo called ckeditor5-build-ie11

I think that you just need to commit and push changes in that repo and all other touched repos. In all cases – you need to do that to your forks of those repos. Plus, you would have to change `mgit.json` so it points out to your forks. Like this:

```
"@ckeditor/ckeditor5-engine": "tuespetre/ckeditor5-engine#branch-name",
```

(you can omit `#branch-name` if you were committing straight to `master`)

Regarding some webpack configs and a sample – you can place those in the main repo (the fork of `ckeditor5`). From that repository, you can import all other pkgs, some polyfills (if you'll install them), etc.

**Comment by tuespetre** [comment]
I'm currently fighting some caret weirdness so I'm learning more about `Selection`/`MutationObserver`/`contenteditable` implementation differences to try and nail it down. I'm mostly comparing between IE11, Edge, and Chrome. I'm seeing some bits of the weirdness surface in Edge so maybe some more good can come out of this endeavor besides just supporting an antiquated browser :trollface:

**Comment by Reinmar** [comment]
> I'm seeing some bits of the weirdness surface in Edge so maybe some more good can come out of this endeavor besides just supporting an antiquated browser

I can feel your pain :D Edge is causing us some headaches – personally, I've got a feeling that it's even less stable than old IE versions ;/. One tip – the F12 dev tools change the bahaviour of the browser, so we try to remember to debug things with the console closed (sic!)

Anyway, 🤞 

If you'll find some interesting Edge bugs (and or workarounds) let us know – it sometimes help to 👍 on the MS's issue tracker.

**Comment by tuespetre** [comment]
I’ve submitted a PR to the engine repository that addresses the thing I was seeing in Edge (and Firefox!) but it will need a little feedback because I’m not sure how to handle the existing test cases surrounding it.

One IE11-only issue I have found is that upon the *user* focusing an editing host via a click, `selectionchange` fires twice: once to place the caret at position 0, and again to place the caret where the user actually intended for it to be by clicking. Meanwhile, upon programmatic `focus`, it sets the position to 0 (say, after inserting a link and the plugin gives focus back to the editor.) If I can find a good mechanism for detecting ‘click focus’ versus ‘programmatic focus’ I should be able to do something there.

The only other one I am seeing yet is a tough one because it needs me to study the workings of the engine itself to figure out. It’s related to `MutationObserver` and the two-step caret binding. I’m seeing cases where whitespace cannot be inserted at the end of a highlighted link, or it is inserted and then ‘breaks off’ when other text is typed, or an `IndexSizeError` is thrown when `_updateDomSelection` tries to modify the selection (due to the DOM text node being shorter than expected as in the model’s text node, for some reason.)

I’ll keep trying to hash these out as separate things to make it easier to integrate them, but I will try and push the `ckeditor5-build-ie11` up before then so maybe others can at least check it out.

**Comment by Reinmar** [comment]
I've seen and replied to your PR (https://github.com/ckeditor/ckeditor5-engine/pull/1442). We still need some work there, but it's promising :)

> Meanwhile, upon programmatic focus, it sets the position to 0 (say, after inserting a link and the plugin gives focus back to the editor.)

I remember there were significant differences between how IE and normal browsers handle selection in editable hosts which lost focus. You'd need to change this scenario behaves in various browsers:

1. Make some selection inside a native `contentEditable` element.
2. Click outside.
3. Focus it via some non-intrusive way (e.g. setting a `setTimeout()` earlier; don't use the console as it affects the focus and selection).

Will other browsers restore the previous selection? Or will all browsers set the new selection (on focus) on position `0`?

If all do the latter, we'd have a bug everywhere (unless `focus` is fired asynchronously). That's because we do this in `editor.editing.view.render()`:

```js
	focus() {
		if ( !this.document.isFocused ) {
			const editable = this.document.selection.editableElement;

			if ( editable ) {
				this.domConverter.focus( editable );
				this.render();
			} else {
				/**
				 * Before focusing view document, selection should be placed inside one of the view's editables.
				 * Normally its selection will be converted from model document (which have default selection), but
				 * when using view document on its own, we need to manually place selection before focusing it.
				 *
				 * @error view-focus-no-selection
				 */
				log.warn( 'view-focus-no-selection: There is no selection in any editable to focus.' );
			}
		}
	}
```

We first focus the editable element and then, synchronously, render the current view. I don't remember now how it works – does `render()` restore the previous selection because `domConverter.focus()` caused asynchronous focus so when `render()` is called we still have the old selection in the model/view? Or are modern browsers restoring the selection by themselves so the fired `focus` event doesn't reset the last stored selection?

Anyway, what I can tell now is that IEs always had problems with the fact that there could be just one selection in the entire window. So, once you took the focus out of the editable, the selection was irreversibly lost there. In the past, in CKEditor 3, we used iframes (so we have multiple windows) and modals (which have focusable input fields) were placed in the main window while the editable was places in an inner window. Then, when we introduced inline editables (in CKEditor 4), we had to introduce selection locking and restoring to handle selection preservation when moving focus to focusable elements in the UI. I don't know now, though, whether this was needed still only for IE or for all browsers. In CKEditor 5 we don't have a selection restoring system yet and everything worked really fine for now. So, how's IE11 different here?

**Comment by Reinmar** [comment]
> The only other one I am seeing yet is a tough one because it needs me to study the workings of the engine itself to figure out. It’s related to MutationObserver and the two-step caret binding. I’m seeing cases where whitespace cannot be inserted at the end of a highlighted link, or it is inserted and then ‘breaks off’ when other text is typed, or an IndexSizeError is thrown when _updateDomSelection tries to modify the selection (due to the DOM text node being shorter than expected as in the model’s text node, for some reason.)

I'd recommend progressive enhancement – does [disabling the two-step caret movement](https://github.com/ckeditor/ckeditor5-link/blob/master/src/linkediting.js#L68) help? If so, let's forget about it in IE11 ;) If not, we'll need to talk about input handling, which is something I wouldn't like to start :D

**Comment by tuespetre** [comment]
I figured out the `IndexSizeError` problem in IE for inserting whitespace at the end of a link. IE has a peculiar behavior with setting the `href` attribute (either via the property or via `setAttribute`) of an `<a>` that is contained within a connected editing host -- it updates the `textContent` to match the value being set for `href` if the `href` and the `textContent` pointed at the same URI. This wreaked havoc for the rendering system. Fortunately it's something I was able to feature test for and shim in the IE11 polyfills file.

```
(function() {
    var host = document.createElement('div');
    host.contentEditable = true;
    host.style.display = 'none';
    var link = document.createElement('a');
    link.textContent = 'https://example.org';
    link.href = 'https://example.org';
    host.appendChild(link);
    document.documentElement.appendChild(host);
    link.href = 'https://example.org?';
    if (link.textContent === 'https://example.org?') {    
        var oldHrefDescriptor = Object.getOwnPropertyDescriptor(HTMLAnchorElement.prototype, 'href');
        Object.defineProperty(HTMLAnchorElement.prototype, 'href', {
            get: oldHrefDescriptor.get,
            set: function(value) {
                var hasElements = false;
                for (var i = 0; i < this.childNodes.length; i++) {
                    if (this.childNodes[i] instanceof Element) {
                        oldHrefDescriptor.set.call(this, value);
                        return;
                    }
                }                
                var oldText = this.textContent;
                oldHrefDescriptor.set.call(this, value);
                if (this.textContent !== oldText) {
                    this.textContent = oldText;
                }
            }
        });
        var oldSetAttribute = Element.prototype.setAttribute;
        Object.defineProperty(HTMLAnchorElement.prototype, 'setAttribute', {
            value: function(key, value) {
                if (key.toLowerCase().trim() === 'href') {
                    this.href = value;
                }
                else {
                    oldSetAttribute.call(this, key, value);
                }
            }
        });
    }
    document.documentElement.removeChild(host);
})();
```

**Comment by tuespetre** [comment]
So the issue I've seen in Edge is that when you key in a space in the middle of a bunch of spaces, it fires extraneous events -- first a mutation observer callback with a `characterData` record at which point the selection has not changed at all (unlike FF, Chrome, and IE, where the selection has changed to be the position after the newly inserted character), then another mutation observer callback with two records (added and removed nodes) at which point the selection has changed to be the position after the first space in the sequence of spaces (*what the hell?*), then finally it fires `input` (totally backwards, that should come before the mutation records) and `selectionchange`, and during both of those events the selection has changed to be after the newly inserted character. But the way the CKEditor engine is responding to these events, it's forcing Edge to keep the caret at the point of insertion, which causes subsequent spaces to be in 'insertion mode'.

After carefully recording events and mutations among the different browsers, it's looking like it's going to be near-impossible to get it all right without... **input handling** 🎉  

The only truly reliable input events I'm seeing are `keydown` and `keyup` as a pair, or `compositionstart` and `compositionend` as a pair -- the browsers handle `keydown` and `keyup` along with `beforeinput`, `input`, and `selectionchange` very differently during composition.

**Comment by Jetski5822** [comment]
Hey guys, any idea when this will be ready?

**Comment by Suraj151** [comment]
i able to transform compiled ckeditor.js (es6) to es5 with babel plugins to support IE11 but now `Symbol` feature give me object error like `TypeError: Unable to get property 'get' of undefined or null reference.` for the all Symbols defined in packages.
any help to solve it as i want it immediately to integrate this editor in our site as soon as possible

**Comment by AntonMedviediev** [comment]
Hello  Guys,
does anyone from you gain worked ES5 version of the ckeditor or this is a trip without end?
I'm ashamed of me, but I really need to make this stuff works on ie11.

As I understand from all of your message it`s really tricky task.
I hope that someone from you will answer how far he went.

Personally, I want to try rewrite it to TS and put ES5 compilation to the awesome-ts-loader, I think it should be simplier than fight with babel, polyfills and so on and so far )

Hope on your wise advices  :)

**Comment by tuespetre** [comment]
Getting it to go to ES5 was just a matter of hunting down a couple/few oddities. Then there was getting the build with CSS going.

Once you’ve got that done though, you still have left the cross-browser issues with the event/selection model. That’s too much work to get right alone, I’m afraid.

**Comment by vitality82** [comment]
Must agree with all about IE, but having clients using IE11 is a deal breaker. Reverting back to v4.

**Comment by Reinmar** [comment]
BTW, I rechecked how long IE11 will be supported.

https://www.microsoft.com/en-us/windowsforbusiness/end-of-ie-support

> Beginning January 12, 2016, only the most current version of Internet Explorer available for a supported operating system will receive technical supports and security updates. Internet Explorer 11 is the last version of Internet Explorer, and will continue to receive security updates, compatibility fixes, and technical support on Windows 7, Windows 8.1, and Windows 10.

According to 
https://support.microsoft.com/en-us/help/13853/windows-lifecycle-fact-sheet

the end of support for Windows 10 looks like a million years from now (2028 and increasing):

![screen shot 2018-11-09 at 17 57 08](https://user-images.githubusercontent.com/545175/48276707-f68cc900-e448-11e8-911e-a819d8e2a75c.png)

Random URLs about the topic:

 * https://medium.com/@burger.neal/the-end-of-life-of-internet-explorer-11-12736f9ff75f
 * https://www.quora.com/When-will-Microsoft-stop-supporting-IE-11-I-cant-find-anything-via-Google

**Comment by brianteeman** [comment]
and sadly that is why ie11 support is requested.

personally speaking I am happy if there is a definitive statement saying "no there won't be an ie11 version" rather than hanging in limbo

**Comment by Reinmar** [comment]
> personally speaking I am happy if there is a definitive statement saying "no there won't be an ie11 version" rather than hanging in limbo

We successfully cooperated with many companies in the past which resulted in delivering many features that might not have been developed otherwise, so I can't say – "it won't be done". It's an open topic. But I understand that a definitive answer would make it easier for you to make a decision :) I'm sorry for not being able to give it.

**Comment by duracell80** [comment]
For enterprise users I can see them sticking with IE11 over Edge for many years due to legacy web apps that were working on IE an that probably won't work on Edge. Some users may have Windows 10 but their browser will be IE11. 

Basically IE6 is back.

Internet Explorer never was evergreen.

**Comment by duracell80** [comment]
We'd have to look at dropping CK Editor / Alloy Editor from our instance of Liferay DXP and move towards TinyMCE if our business direction required continued usage in IE. We'd have to move on to other solutions that weren't CK Editor.

**Comment by duracell80** [comment]
Fool me once shame on you, fool me 7 times shame on all of us. Just how evergreen is edge https://www.construct.net/en/blogs/construct-official-blog-1/just-how-evergreen-is-microsoft-edge-854

**Comment by wwalc** [comment]
> We'd have to look at dropping CK Editor / Alloy Editor from our instance of Liferay DXP and move towards TinyMCE if our business direction required continued usage in IE. We'd have to move on to other solutions that weren't CK Editor.

@duracell80 Technically speaking, TinyMCE (even their latest "v5") is having a similar architecture as CKEditor 4. So, personally, I wouldn't see any reasonable benefit for doing it. Keep in mind that CKEditor 4 will be maintained still for several years (until 2023) - we have lots of users & customers using it, so we'd not abandon it just like that. As long as IE11 is for you a priority (or even... IE8) you may continue using CKEditor 4 / Alloy Editor and just keep an eye on the status of IE11.

**Comment by jacekbogdanski** [comment]
Hello @prachh,

currently, we are in progress of developing Angular integration for CKEditor4. You can subscribe https://github.com/ckeditor/ckeditor-dev/issues/2481 for updates on its state. We are planning to keep CKEditor4 browser support, so it will be probably only limited to the [Angular browser support](https://angular.io/guide/browser-support) where IE9-11 is supported.

Be aware that Angular integration will land in a separate repository. No worries, we will add information about the location in the mentioned GH ticket reference.

**Comment by prachh** [comment]
> Hello @prachh,
> 
> currently, we are in progress of developing Angular integration for CKEditor4. You can subscribe [ckeditor/ckeditor-dev#2481](https://github.com/ckeditor/ckeditor-dev/issues/2481) for updates on its state. We are planning to keep CKEditor4 browser support, so it will be only limited by [Angular browser support](https://angular.io/guide/browser-support) where IE9-11 is supported.
> 
> Be aware that Angular integration will land in a separate repository. No worries, we will add information about the location in the mentioned GH ticket reference. If you would like to get more information about upcoming integration, feel free to ask questions in [ckeditor/ckeditor-dev#2481](https://github.com/ckeditor/ckeditor-dev/issues/2481) ticket - our discussion may be valuable for the CKEditor4 community.

Thanks @jacekbogdanski

**Comment by runata** [comment]
I am working on the ckeditor compatibility issue right now. After building editor for ES5 I got to the point mentioned by @Suraj151.

> i able to transform compiled ckeditor.js (es6) to es5 with babel plugins to support IE11 but now `Symbol` feature give me object error like `TypeError: Unable to get property 'get' of undefined or null reference.` for the all Symbols defined in packages.
> any help to solve it as i want it immediately to integrate this editor in our site as soon as possible

It seems that I stuck on it for a while, so I would appreciate any advice.

**Comment by webloem** [comment]
I got to the same point a while ago and simply gave up. 

Anyone reading this: please do not make the same mistake I did: transpiling to ES5 will not magically give you IE11 support. CKEditor5 relies on browser APIs that are different or not available in IE.

**Comment by cpeker** [comment]
Any updates regarding this issue?

**Comment by scofalik** [comment]
Hello everyone,

we concluded the research about running CKEditor 5 on IE 11. In overall it is possible to provide a CKE5 build for IE 11 and after using some polyfills and quick fixes it loads and base features work (like typing). Collaboration features seem to be working too. Unfortunately, without further work, the editor is not stable enough to be used on production. Some of the important things that do not work:

1. Clipboard uses different API and needs a special workaround to support pasting HTML (so-called “pastebin”). As of now, the editor crashes when using the clipboard.
2. There are a few problems with selection (some are discussed in this thread).
3. There are a few smaller glitches, bugs or crashes, for example, balloon toolbars are not positioned at all.
4. Performance is an issue. It is not in an unusable territory but there’s a visible difference in how smooth typing is (compared to other browsers).
5. Widgets (mostly tables) need some fixing/improvements.

And these are only some things that would need to be done.

The research was promising in the sense that it proved that support for IE11 is feasible. However, it will require a significant effort to become anywhere close to “production ready” and we’re not currently able to fund it ourselves. Therefore, we don’t have any plans yet regarding any ETA as we’re still looking for company(ies) who be willing to share the cost of this project.

If you want to play with CKE5 in IE 11, here is an instruction on how to create a build for IE 11. Only the classic editor was tested and available, although you could probably prepare other editors looking at changes done in `ckeditor5-build-classic`.

1. Clone `ckeditor5` repository. Checkout to `ie11` branch.
2. Run `mgit sync`. This one may fail on Windows, run it until it will pass without errors.
3. Run `yarn install`.
4. Go to `packages/ckeditor5-build-classic`.
5. Run `yarn build`.
6. `ckeditor.js` file will be generated in `ckeditor5-build-classic/build`.

If you are interested in what exactly has been changed, check which repositories are switched to `ie11` branch and use whatever tool you like to see changes introduced in commits. To check which branches are switched to `ie11` run `mgit st` in the `ckeditor5` repository main directory.

**If you’d like to pick this up and help with fixing issues on IE 11, you are most welcome to do so!**

**Comment by mleathem** [comment]
(Thanks for you work on this... sadly IE11 is a "must have", _so I'm told_ ... )

To be honest, I'm struggling with this:    got so far as...
$ yarn run mgit sync  
- but keeps failing : 
> "Host key verification failed"  
> "fatal: could not read from the remote repository"
-- 
I've tried a few diffrent clone (with my github username - thinking maybe this was the credential fail)
even if  /package/ dirs are being created, (from cloing the node_modules) but the dirs are empty.

--- background  - If it helps anyone: 
the ticket here doesnt mention  **mgit sync**  must relate to https://www.npmjs.com/package/mgit  - so inststall globally.
then:    $ npm run bootstrap  
( this runs:  mgit sync && yarn install  )
but that's as far as I get because:     $ yarn run mgit sync   just returns "Host key verification failed"

**Comment by scofalik** [comment]
Ooops!

`mgit` is our [own project which is actually named mgit2](https://www.npmjs.com/package/mgit2) (though the command is `mgit`). Sorry about missing that one. It should be installed after you run `yarn install` as it is listed in `package.json`. However, I am not sure if it doesn't conflict if you have `mgit` installed. [You can also try installing mgit2 globally](https://ckeditor.com/docs/ckeditor5/11.1.1/framework/guides/contributing/development-environment.html#setting-up-the-ckeditor-development-environment), using npm or yarn.

BTW. Why do you run `yarn run mgit sync`?

**Comment by Yahav** [comment]
So, can we get a status on IE11 support? where does it stand? did you guys ever brought it to a stable, production worthy state?

**Comment by wwalc** [comment]
@Yahav So far no companies were interested in sponsoring the work on IE11 compatibility - this is a blocker for us to move this issue forward because IE11 support requires a lot of work. So for now, we gave up after closing the research that @scofalik mentioned in [his comment](https://github.com/ckeditor/ckeditor5/issues/330#issuecomment-485749345).

**Comment by coolswing** [comment]
Hi,
on ie11 i get the error : SCRIPT1003: Expected ':'  when including @ckeditor/ckeditor5-vue to my vue-cli application making it crashed.

**Comment by Rialzista** [comment]
Hello @wwalc , will IE11 be ever supported?.
Are we safe to just use CKeditor 4 for the next 5 years? 
Or would we be required to upgrade?

**Comment by wwalc** [comment]
@Rialzista There is no progress regarding IE11 (by not spending time on it, we're able to move on and continue adding more new features: https://ckeditor.com/blog/CKEditor-5-v15.0.0-with-horizontal-line-page-break-responsive-toolbar-and-SVG-upload-support/)

As for CKEditor 4: the official support for CKEditor 4 is until 2023. This is something that we committed to already, and for sure we will not end up the support earlier than that. Which gives you 4 years of peace of mind. Closer to that date we'd have a look at how the situation looks like and decide what next with CKEditor 4.

**Comment by florianchevallier** [comment]
Hi,

With the build from [this post](https://github.com/ckeditor/ckeditor5/issues/330#issuecomment-485749345), do we get something like the one build from the CDN ? I get this error when using the file built by yarn : 

```
Uncaught CKEditorError: ckeditor-duplicated-modules: Some CKEditor 5 modules are duplicated. Read more:
```

**Comment by techboyg5** [comment]
I don't think that this will happen, as Internet Explorer 11 doesn't support ES6.

**Comment by godai78** [comment]
This issue is no longer relevant as IE11 reaches the end of support as of June 2022.
